### PR TITLE
batches: Isolate `has_webhooks` oobmigration

### PR DESCRIPTION
**Pull request by efritz**
This PR *isolates* oobmigration #13 so that it does not need any _behavioral_ imports that can change over time. This is in preparation for its deprecation, where it will run only as part of the migrator for historic upgrades. We want to *freeze* this migration as much as possible, because we want to treat it with consistent behavior over time to allow the migrator to make some assumptions (like the registered code for an oobmigration is valid at any point in its lifetime).

This PR makes progress towards #38050, but still needs to be fed configuration keys from the worker. A follow-up PR will take care of this.

Feel free to review by-commit which should make the refactoring very obvious.

## Test plan

Existing unit tests.

### CODE REVIEW

File: internal/oobmigration/migrations/extsvc_webhook_migrator.go [hunk 1]
```diff
@@ -23,65 +22,150 @@ type ExternalServiceWebhookMigrator struct {
...
+		defer func() { err = basestore.CloseRows(rows, err) }()
+
+		for rows.Next() {
+			var id int
+			var kind, config, keyID string
+			if err := rows.Scan(&id, &kind, &config, &keyID); err != nil {
+				return nil, err
+			}
+			if keyID != "" {
+				panic("UNSUPPORTED") // TODO
```
Comment (efritz) [inline internal/oobmigration/migrations/extsvc_webhook_migrator.go:92]: We need to figure out how to support encryption keys here.

**Comment by efritz** [comment]
@sourcegraph/batchers Would you like to deprecate this oobmigration? This would allow you to make the column `has_webhooks` on the `external_services` table non-nullable.

**Comment by eseliger** [comment]
I think so, yes. @LawnGnome can you confirm that'd be safe?

**Comment by LawnGnome** [comment]
I think so, yes. I actually expect that we'll want to remove this entirely in 4.0.

**Comment by efritz** [comment]
Follow-up PRs will include isolation of the tests and supplying encryption keys in a uniform way.

File: internal/oobmigration/migrations/extsvc_webhook_migrator.go [hunk 2]
```diff
@@ -2,86 +2,179 @@ package migrations
...
+			}
+			config, err = encryption.MaybeDecrypt(ctx, m.key, config, keyID)
+			if err != nil {
+				return nil, err
+			}
 
-	svcs, err := store.List(ctx, database.ExternalServicesListOptions{
-		OrderByDirection: "ASC",
-		LimitOffset:      &database.LimitOffset{Limit: m.BatchSize},
-		NoCachedWebhooks: true,
```
Comment (LawnGnome) [inline internal/oobmigration/migrations/extsvc_webhook_migrator.go:72]: Filed #40219 to remind myself to remove this option once this lands.

**Review by LawnGnome — APPROVED** [review]
LGTM, although again, I would consider how much follow up work you want to do if we remove this in a month anyway.

**Comment by efritz** [comment]
> LGTM, although again, I would consider how much follow up work you want to do if we remove this in a month anyway.

@LawnGnome FYI this is in preparation to deprecate _and_ "freeze" the oobmigration. The current direction I'm going with multi-version upgrades require us to keep the code around until some sort of EOL period. That's why I'm minimizing dependencies on anything that could change over time (saving db connection stuff and query utilities).

**Pull request merged**